aptos-move/move-examples/defi/aptoswap/core/sources/liquidity_pool.move

```diff
@@ -0,0 +1,417 @@
+module pool_manager::liquidity_pool {
+    use aptos_framework::fungible_asset::{Metadata, FungibleStore, FungibleAsset, deposit};
+    use aptos_framework::object::{Object, ExtendRef, generate_signer_for_extending, generate_extend_ref};
+    use aptos_framework::fungible_asset;
+    use aptos_framework::object;
+    use aptos_framework::primary_fungible_store;
+    use aptos_std::comparator;
+    use aptos_std::math128;
+    use aptos_std::math64;
+    use aptos_std::smart_table::SmartTable;
+    use aptos_std::smart_table;
+    use std::bcs;
+    use std::error;
+    use std::option;
+    use std::string;
+    use std::vector;
+
+    use pool_manager::utilities;
+
+    const MINIMAL_LIQUIDITY: u64 = 1000;
+
+    // fee is basis point based so the scale is 10000.
+    const FEE_SCALE: u64 = 10000;
+
+    const ERR_SWAP_PAIR_CANNOT_BE_THE_SAME: u64 = 1;
+    const ERR_POOL_EXISTENCE: u64 = 2;
+    const ERR_NOT_ENOUGH_LIQUIDITY: u64 = 3;
+    const ERR_NOT_ENOUGH_LP_TOKEN_TO_BURN: u64 = 4;
+    const ERR_ZERO_COIN_TO_SWAP: u64 = 5;
+    const ERR_INCORRECT_SWAP_DUE_TO_REDUCED_LIQUIDITY: u64 = 6;
+
+    /// Stored at lp token metadata object to find the liquidity pool.
+    /// x address must be lexicographically smaller than y address.
+    struct SwapPair has copy, drop, key {
+        x: Object<Metadata>,
+        y: Object<Metadata>,
+    }
+
+    struct LiquidityPool has key {
+        extend_ref: ExtendRef,
+        store_pair: vector<Object<FungibleStore>>,
+        lp_token_mint_ref: fungible_asset::MintRef,
+        lp_token_burn_ref: fungible_asset::BurnRef,
+        fee: u64,
+        // basis points
+    }
+
+    /// The onchain index of the liquidity pools of different pairs
+    struct LiquidityPoolManager has key {
+        extend_ref: ExtendRef,
+        map: SmartTable<vector<Object<Metadata>>, Object<LiquidityPool>>
+    }
+
+    /// Intialize the module in pool manager object
+    fun init_module(admin: &signer) {
+        let extend_ref = aptoswap::init::retrieve_extend_ref(admin);
+        let pool_manager_obj_signer = object::generate_signer_for_extending(&extend_ref);
+        move_to(&pool_manager_obj_signer, LiquidityPoolManager {
+            extend_ref,
+            map: smart_table::new()
+        })
+    }
+
+    /// Register liquidity pool `X`/`Y`.
+    public fun register(acc: &signer, x: Object<Metadata>, y: Object<Metadata>) acquires LiquidityPoolManager {
+        if (!is_valid_pair(x, y)) {
+            return register(acc, y, x)
+        };
+        assert!(!lp_exists(x, y), error::already_exists(ERR_POOL_EXISTENCE));
+
+        let (lp_token_name, lp_token_symbol) = utilities::swap_lp_token_name_and_symbol(x, y);
+
+        let pool_manager_signer = &get_pool_account_signer();
+        let pool_cref = &object::create_object_from_account(pool_manager_signer);
```

Comment:
i wonder when should we choose:
- create_object_from_account
- create_named_object
- create_object_from_object
?

Reply:
1. create_object_from_account -> your new object owner is an account.
2. create_named_object -> your new object need deterministic naming, such as primary fungible store, or cannot be deleted, such as metadata object based on which we issue fungible assets.
3. create_object_from_object -> similar to 1, but the owner now becomes an account.
1 or 3 doesn't matter because you can always create an object from either can transfer the ownership. So the comparison is actually `named` vs `randomly generated`

Reply:
@lightmark why does object ownership matter - i guess that's a follow-up question

Reply:
let's assume you have a `FungibleStore` object, owner has the authority to withdraw. That's why ownership matters. you can build the permission model in your way, but the defacto pattern is the ownership model, at least for fungible asset store.